src/memory.rs

```diff
@@ -206,57 +218,54 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
         let id = self.next_id;
         self.next_id.0 += 1;
         self.alloc_map.insert(id, alloc);
         Ok(MemoryPointer::new(id, 0))
     }
 
-    // TODO(solson): Track which allocations were returned from __rust_allocate and report an error
-    // when reallocating/deallocating any others.
-    pub fn reallocate(&mut self, ptr: MemoryPointer, old_size: u64, old_align: u64, new_size: u64, new_align: u64) -> EvalResult<'tcx, MemoryPointer> {
+    pub fn reallocate(&mut self, ptr: MemoryPointer, old_size: u64, old_align: u64, new_size: u64, new_align: u64, kind: Kind) -> EvalResult<'tcx, MemoryPointer> {
         use std::cmp::min;
 
-        // TODO(solson): Report error about non-__rust_allocate'd pointer.
         if ptr.offset != 0 || self.get(ptr.alloc_id).is_err() {
             return Err(EvalError::ReallocateNonBasePtr);
         }
-        if self.get(ptr.alloc_id).ok().map_or(false, |alloc| alloc.static_kind != StaticKind::NotStatic) {
-            return Err(EvalError::ReallocatedStaticMemory);
+        if let Ok(alloc) = self.get(ptr.alloc_id) {
+            if alloc.kind != kind {
+                return Err(EvalError::ReallocatedWrongMemoryKind(alloc.kind, kind));
+            }
         }
 
         // For simplicities' sake, we implement reallocate as "alloc, copy, dealloc"
-        let new_ptr = self.allocate(new_size, new_align)?;
+        let new_ptr = self.allocate(new_size, new_align, kind)?;
         self.copy(ptr.into(), new_ptr.into(), min(old_size, new_size), min(old_align, new_align), /*nonoverlapping*/true)?;
-        self.deallocate(ptr, Some((old_size, old_align)))?;
+        self.deallocate(ptr, Some((old_size, old_align)), kind)?;
 
         Ok(new_ptr)
     }
 
-    // TODO(solson): See comment on `reallocate`.
-    pub fn deallocate(&mut self, ptr: MemoryPointer, size_and_align: Option<(u64, u64)>) -> EvalResult<'tcx> {
+    pub fn deallocate(&mut self, ptr: MemoryPointer, size_and_align: Option<(u64, u64)>, kind: Kind) -> EvalResult<'tcx> {
         if ptr.offset != 0 || self.get(ptr.alloc_id).is_err() {
-            // TODO(solson): Report error about non-__rust_allocate'd pointer.
             return Err(EvalError::DeallocateNonBasePtr);
         }
 
-        {
-            // deallocate_local in eval_context.rs relies on nothing actually having changed when this error occurs.
-            // So we do this test in advance.
-            let alloc = self.get(ptr.alloc_id)?;
-            if alloc.static_kind != StaticKind::NotStatic {
-                return Err(EvalError::DeallocatedStaticMemory);
-            }
-            if let Some((size, align)) = size_and_align {
-                if size != alloc.bytes.len() as u64 || align != alloc.align {
-                    return Err(EvalError::IncorrectAllocationInformation);
-                }
+        let alloc = match self.alloc_map.remove(&ptr.alloc_id) {
+            Some(alloc) => alloc,
+            None => return Err(EvalError::DoubleFree),
```

Comment:
Actually I just noticed this line is unreachable -- we already check above `self.get(ptr.alloc_id).is_err()`. (There a similar redundant conditional in `reallocate`.)